File: scripts/bootstrap_admin.ts
#!/usr/bin/env tsx
/**
 * PAHAMI V2 — Super Admin Bootstrap CLI Script
 * Creates or resets the initial Super Administrator with scrypt password hashing.
 * Usage: npx tsx scripts/bootstrap_admin.ts --username <user> --password <pass> --name "Full Name"
 */

import crypto from "crypto";
import { hashPasswordSync } from "../lib/admin/crypto";
import { adminRepository } from "../lib/admin/admin-repository";

function parseArgs() {
  const args = process.argv.slice(2);
  const params: Record<string, string> = {};

  for (let i = 0; i < args.length; i++) {
    if (args[i].startsWith("--")) {
      const key = args[i].replace(/^--/, "");
      const val = args[i + 1] && !args[i + 1].startsWith("--") ? args[i + 1] : "true";
      params[key] = val;
    }
  }

  return params;
}

function validatePassword(pass: string): { valid: boolean; reason?: string } {
  if (pass.length < 8) {
    return { valid: false, reason: "Password minimal 8 karakter." };
  }
  if (!/[A-Z]/.test(pass)) {
    return { valid: false, reason: "Password harus mengandung minimal 1 huruf besar (A-Z)." };
  }
  if (!/[0-9]/.test(pass)) {
    return { valid: false, reason: "Password harus mengandung minimal 1 angka (0-9)." };
  }
  return { valid: true };
}

async function main() {
  console.log("==================================================================");
  console.log("PAHAMI V2 — SUPER ADMIN BOOTSTRAP WIZARD");
  console.log("==================================================================");

  const args = parseArgs();
  const username = args.username || "irham_admin";
  const password = args.password || "PahamiMadiun2026!";
  const fullName = args.name || "Irham Najib (Super Admin Developer)";

  console.log(`Username Target : ${username}`);
  console.log(`Nama Lengkap    : ${fullName}`);

  const validation = validatePassword(password);
  if (!validation.valid) {
    console.error(`[GAGAL] Password tidak memenuhi standar keamanan: ${validation.reason}`);
    process.exit(1);
  }

  const { hash, salt } = hashPasswordSync(password);

  // Check if admin already exists
  const existing = adminRepository.getAdminByUsername(username);
  if (existing) {
    console.log(`[INFO] Admin '${username}' sudah ada. Memperbarui password dan memastikan role SUPER_ADMIN...`);
    existing.password_hash = hash;
    existing.salt = salt;
    existing.role = "SUPER_ADMIN";
    existing.is_active = true;
    existing.failed_login_attempts = 0;
    existing.locked_until = null;
    existing.updated_at = new Date().toISOString();
  } else {
    console.log(`[INFO] Membuat akun Super Admin baru '${username}'...`);
    adminRepository.createAdmin({
      username,
      password_hash: hash,
      salt,
      full_name: fullName,
      role: "SUPER_ADMIN",
      is_active: true,
      failed_login_attempts: 0,
      locked_until: null,
      last_login_at: null,
    });
  }

  adminRepository.recordAuditLog({
    admin_username: username,
    action: "SUPER_ADMIN_BOOTSTRAPPED",
    target_type: "ADMIN",
    target_id: username,
    result: "SUCCESS",
    metadata: { via: "cli_bootstrap", scrypt_n: 16384 },
  });

  console.log("==================================================================");
  console.log("[SUKSES] Akun Super Admin berhasil dikonfigurasi!");
  console.log(`- Username: ${username}`);
  console.log(`- Role    : SUPER_ADMIN (Full Permissions)`);
  console.log(`- Status  : Aktif`);
  console.log("Gunakan kredensial ini untuk login di: http://localhost:3000/admin");
  console.log("==================================================================");
}

main().catch((err) => {
  console.error("[ERROR] Bootstrap gagal:", err);
  process.exit(1);
});

File: scripts/bootstrap_admin.py
#!/usr/bin/env python3
"""
PAHAMI V2 — Super Admin Bootstrap Python Script
Directly inserts or updates a Super Administrator in the remote Supabase PostgreSQL database.
Usage: uv run --with "psycopg[binary]" python scripts/bootstrap_admin.py --username irham_admin --password PahamiMadiun2026!
"""

import os
import sys
import re
import hashlib
import secrets
import argparse
from pathlib import Path

try:
    import psycopg
except ImportError:
    print("[ERROR] 'psycopg' library is required.")
    print("Run with: uv run --with \"psycopg[binary]\" python scripts/bootstrap_admin.py")
    sys.exit(1)


def load_env_local(project_root: Path) -> dict:
    env_file = project_root / ".env.local"
    env_vars = {}
    if env_file.exists():
        with open(env_file, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                match = re.match(r"^([A-Za-z0-9_]+)=(.*)$", line)
                if match:
                    key = match.group(1).strip()
                    val = match.group(2).strip().strip("\"'")
                    env_vars[key] = val
    return env_vars


def hash_password_scrypt(password: str, salt: str = None) -> tuple[str, str]:
    if not salt:
        salt = secrets.token_hex(16)
    # scrypt with N=16384, r=8, p=1, key_length=64 bytes
    derived = hashlib.scrypt(
        password.encode("utf-8"),
        salt=salt.encode("utf-8"),
        n=16384,
        r=8,
        p=1,
        maxmem=0,
        dklen=64
    )
    return derived.hex(), salt


def main():
    parser = argparse.ArgumentParser(description="Bootstrap Super Admin for PAHAMI V2")
    parser.add_argument("--username", default="irham_admin", help="Admin username")
    parser.add_argument("--password", default="PahamiMadiun2026!", help="Admin password (min 8 chars, 1 uppercase, 1 digit)")
    parser.add_argument("--name", default="Irham Najib (Super Admin Developer)", help="Admin full name")
    args = parser.parse_args()

    project_root = Path(__file__).resolve().parent.parent
    env_vars = load_env_local(project_root)
    db_url = os.getenv("DATABASE_URL") or env_vars.get("DATABASE_URL")

    if not db_url:
        print("[ERROR] DATABASE_URL tidak ditemukan di .env.local!")
        sys.exit(1)

    print("==================================================================")
    print("PAHAMI V2 — REMOTE SUPABASE SUPER ADMIN BOOTSTRAP")
    print("==================================================================")
    print(f"Target Username: {args.username}")
    print(f"Target Name    : {args.name}")

    if len(args.password) < 8 or not any(c.isupper() for c in args.password) or not any(c.isdigit() for c in args.password):
        print("[ERROR] Password minimal 8 karakter, harus mengandung huruf besar dan angka.")
        sys.exit(1)

    pwd_hash, salt = hash_password_scrypt(args.password)

    try:
        conn = psycopg.connect(db_url, autocommit=True)
        cursor = conn.cursor()

        sql = """
        INSERT INTO admin_accounts (username, password_hash, salt, full_name, role, is_active)
        VALUES (%s, %s, %s, %s, 'SUPER_ADMIN', true)
        ON CONFLICT (username) DO UPDATE
        SET password_hash = EXCLUDED.password_hash,
            salt = EXCLUDED.salt,
            full_name = EXCLUDED.full_name,
            role = 'SUPER_ADMIN',
            is_active = true,
            failed_login_attempts = 0,
            locked_until = NULL,
            updated_at = now();
        """
        cursor.execute(sql, (args.username, pwd_hash, salt, args.name))

        # Record audit log
        audit_sql = """
        INSERT INTO admin_audit_logs (admin_username, action, target_type, target_id, result, metadata)
        VALUES (%s, 'SUPER_ADMIN_BOOTSTRAPPED', 'ADMIN', %s, 'SUCCESS', '{"via": "python_cli"}'::jsonb);
        """
        cursor.execute(audit_sql, (args.username, args.username))

        print("[SUKSES] Akun Super Admin berhasil dibuat/diperbarui di remote Supabase!")
        print(f"- Username : {args.username}")
        print(f"- Role     : SUPER_ADMIN")
        print("Silakan login melalui: http://localhost:3000/admin")
        print("==================================================================")
        conn.close()
    except Exception as e:
        print(f"[ERROR] Database execution error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
